model/map-ldap/src/main/java/org/keycloak/models/map/storage/ldap/role/LdapRoleMapKeycloakTransaction.java

```diff
@@ -0,0 +1,441 @@
+                client.setSingleAttribute("ou", mapped.getClientId());
+                identityStore.add(client);
+
+                tasksOnRollback.add(new DeleteOperation() {
+                    @Override
+                    public void execute() {
+                        identityStore.remove(client);
+                    }
+                });
+
+                // retry creation of client role
+                identityStore.add(mapped.getLdapMapObject());
+            }
+        }
+
+        entities.put(mapped.getId(), mapped);
+
+        tasksOnRollback.add(new DeleteOperation() {
+            @Override
+            public void execute() {
+                identityStore.remove(mapped.getLdapMapObject());
+                entities.remove(mapped.getId());
+            }
+        });
+
+        return mapped;
+    }
+
+    @Override
+    public boolean delete(String key) {
+        LdapMapRoleEntityFieldDelegate read = read(key);
+        if (read == null) {
+            throw new ModelException("unable to read entity with key " + key);
+        }
+        deletedKeys.add(key);
+        tasksOnCommit.add(new DeleteOperation() {
+            @Override
+            public void execute() {
+                LdapMapIdentityStore identityStore = new LdapMapIdentityStore(session, ldapMapConfig);
+                identityStore.remove(read.getLdapMapObject());
+            }
+        });
+        return true;
+    }
+
+    public LdapRoleEntity readLdap(String key) {
+        LdapMapRoleEntityFieldDelegate read = read(key);
+        if (read == null) {
+            return null;
+        } else {
+            return read.getEntityFieldDelegate();
+        }
+    }
+
+    @Override
+    public LdapMapRoleEntityFieldDelegate read(String key) {
+        if (deletedKeys.contains(key)) {
+            return null;
+        }
+
+        // reuse an existing live entity
+        LdapMapRoleEntityFieldDelegate val = entities.get(key);
+
+        if (val == null) {
+
+            // try to look it up as a realm role
+            val = lookupEntityById(key, null);
+
+            if (val == null) {
+                // try to look it up using a client role
+                // currently the API doesn't allow to get a list of all keys, therefore we need a separate attribute
+                // also, getArray is broken as it doesn't look up the parent's values if an entry is empty
+                String clientsToSearch = config.scope("clients").get("clientsToSearch");
+                if (clientsToSearch != null) {
+                    String[] clientIds = clientsToSearch.split("\\s*,\\s*");
+                    for (String clientId : clientIds) {
+                        val = lookupEntityById(key, clientId);
+                        if (val != null) {
+                            break;
+                        }
+                    }
+                }
+            }
```

Comment:
How would the `clients` config be maintained in runtime? Who would be responsible for maintaining the `clientsToSearch` list? How would it react to changes in additions / removals of client ?

Side note: transaction should get a pre-processed config rather than using a `Config.scope()` methods - those are high-level ones and belong to `ProviderFactory`.

Reply:
The clientsToSearch list is here as not all configurations allowed to iterate over their keys via `getPropertyNames()`. It needs to be maintained in sync with the client nodes, what is not ideal. An improvement is IMHO outside of the scope of this issue.

I'll keep the side note in mind for future PRs in this area.

Reply:
The Role Mapper for LDAP is existing functionality AFAIK, see this excerpt from the docs. It works for both realm and client roles. https://www.keycloak.org/docs/latest/server_admin/#_ldap_mappers

> Role Mapper
This mapper configures role mappings from LDAP into Keycloak role mappings. A single role mapper can map LDAP roles (usually groups from a particular branch of the LDAP tree) into roles corresponding to a specified client’s realm roles or client roles. You can configure more Role mappers for the same LDAP provider. For example, you can specify that role mappings from groups under ou=main,dc=example,dc=org map to realm role mappings, and role mappings from groups under ou=finance,dc=example,dc=org map to client role mappings of client finance.

![image](https://user-images.githubusercontent.com/3957921/156525760-da6ab0ce-e1c4-4d4e-965c-8187e2e06cf2.png)

As discussed today, this was intended to be able to specify a DN per client that would be non-numeric. 

This functionality will be removed as you pointed out that the same thing can be achieved using LDAP aliasing.

Reply:
implemented, please re-review.

Reply:
Thank you for the reference. We will deal with this case later on if required by the community.